python/ray/_private/ray_logging.py

```diff
@@ -253,3 +261,153 @@ def emit(self, data):
+        """
+        global _log_dedup_warned
+        if not RAY_DEDUP_LOGS:
+            return lines
+
+        source = (metadata.get("ip"), metadata.get("pid"))
+        now = time.time()
+        output = []
+
+        # Decide which lines to emit and which to buffer.
+        for line in lines:
+            if RAY_TQDM_MAGIC in line:
+                output.append(line)
+                continue
+            dedup_key = _canonicalise_log_line(line)
+            if dedup_key in self.recent:
+                sources = self.recent[dedup_key].sources
+                sources.add(source)
+                if len(sources) > 1:
+                    state = self.recent[dedup_key]
+                    self.recent[dedup_key] = DedupState(
+                        state.timestamp,
+                        state.count + 1,
+                        line,
+                        metadata,
+                        sources,
+                    )
+                else:
+                    # Don't dedup messages from the same source, just print.
+                    output.append(line)
+            else:
+                self.recent[dedup_key] = DedupState(now, 0, line, metadata, {source})
+                output.append(line)
+
+        # Flush patterns from the buffer that are older than the aggregation window.
+        while self.recent:
+            if now - next(iter(self.recent.values())).timestamp > AGGREGATION_WINDOW_S:
+                dedup_key = next(iter(self.recent))
+                state = self.recent.pop(dedup_key)
+                # we already logged an instance of this line immediately when received,
+                # so don't log for count == 0
+                if state.count > 1:
+                    # (Actor pid=xxxx) [repeated 2x across cluster] ...
+                    output.append(
+                        state.line
+                        + self._color(
+                            f" [repeated {state.count}x across cluster]"
+                            + self._warn_once()
+                        )
+                    )
+                    # Continue aggregating for this key but reset timestamp and count.
+                    state.timestamp = now
+                    state.count = 0
+                    self.recent[dedup_key] = state
+                elif state.count > 0:
+                    # Aggregation wasn't fruitful, print the line and stop aggregating.
+                    output.append(line)
+            else:
+                break
+
+        return output
+
+    def flush(self) -> List[dict]:
+        """Return all buffered log messages and clear the buffer.
+
+        Returns:
+            List of log batches to print.
+        """
+        output = []
+        for state in self.recent.values():
+            if state.count > 1:
+                output.append(
+                    dict(
+                        state.metadata,
+                        **{
+                            "lines": [
```

Comment:
This logic seems to be the same as one inside `deduplicate`. Should we take it out to a helper method?